### PR TITLE
Add MLFloat16 support for SoftmaxCrossEntropyLoss for CUDA EP

### PR DESCRIPTION
**Pull request by ashbhandare**
This PR adds support for float16 type for SoftmaxCrossEntropyLoss and SoftmaxCrossEntropyLossGrad which is required for some Huggingface models. 

This PR needs the ONNX commit to be advanced to include the function body fix for NLLLoss and SCELoss
https://github.com/onnx/onnx/pull/3486 
https://github.com/onnx/onnx/pull/3487

### CODE REVIEW

File: orttraining/orttraining/training_ops/cuda/loss/softmax_cross_entropy_loss_impl.cu [hunk 1]
```diff
@@ -221,6 +223,7 @@ SPECIALIZED_IMPL_SoftMaxEntropyLossGradImpl(float, int64_t)
 
 SPECIALIZED_IMPL_ComputeWeightsSoftmaxCrossEntropyImpl(float, int32_t)
 SPECIALIZED_IMPL_ComputeWeightsSoftmaxCrossEntropyImpl(float, int64_t)
+SPECIALIZED_IMPL_ComputeWeightsSoftmaxCrossEntropyImpl(half, int64_t)
```
Comment (weixingzhang) [inline orttraining/orttraining/training_ops/cuda/loss/softmax_cross_entropy_loss_impl.cu:226]: Historically, "specialized" was used here. Actually, it is to instantiate the template function. Can you help change to INSTANTIATE_XXX?

File: orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc [hunk 2]
```diff
@@ -354,6 +407,24 @@ TEST(CudaKernelTest, SoftmaxCrossEntropyLoss_SmallSizeTensor) {
   TestSoftmaxCrossEntropyLoss(&X_dims, &index_dims, nullptr, &Y_dims_none, &log_prob_dims, "none");
 }
 
+#if USE_CUDA
```
Comment (weixingzhang) [inline orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc:410]: Ditto.

File: orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc [hunk 3]
```diff
@@ -369,6 +440,24 @@ TEST(CudaKernelTest, SoftmaxCrossEntropyLoss_MediumSizeTensor) {
...
```
Comment (weixingzhang) [inline orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc:443]: Ditto.

File: orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc [hunk 4]
```diff
@@ -339,6 +366,32 @@ TEST(CudaKernelTest, SoftmaxCrossEntropyLoss_TinySizeTensor) {
   TestSoftmaxCrossEntropyLoss(&X_dims, &index_dims, nullptr, &Y_dims_none, &log_prob_dims, "none", 0);
 }
 
+#if USE_CUDA
```
Comment (weixingzhang) [inline orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc:369]: Please add USE_ROCM as well. Thanks!

File: orttraining/orttraining/training_ops/cuda/loss/softmax_cross_entropy_loss_impl.cc [hunk 5]
```diff
@@ -157,13 +165,26 @@ Status SoftmaxCrossEntropyLoss<T, Tin>::ComputeInternal(OpKernelContext* ctx) co
...
+      ORT_RETURN_IF_ERROR(reduce_sum(
+          Stream(),
+          reinterpret_cast<CudaT*>(tmp_loss_sample_buffer),
+          normalize_factor_data.get(),
+          static_cast<int>(N_D),
+          reduction_buffer.get(),
+          buffer_size));
+      // Convert from accumulation type to output type
+      // Need a better way to do this cast as this is a single element cast
+      Impl_Cast<TBuf, CudaT>(Stream(), normalize_factor_data.get(), reinterpret_cast<CudaT*>(total_loss_data), 1);
```
Comment (weixingzhang) [inline orttraining/orttraining/training_ops/cuda/loss/softmax_cross_entropy_loss_impl.cc:178]: How about extending reduce_sum to support this case rather than using Impl_Cast here?

**Review by weixingzhang — APPROVED** [review]

**Review by codemzs — DISMISSED** [review]
@ashbhandare Thanks for the change, I just reviewed the SoftmaxCrossEntropyLoss code for adding MLFloat16 type, it looks fine but just for sanity sake I would quickly benchmark this kernel in isolation in a loop to make sure there are no perf regression due to casting that comes with this change.

**Comment by ashbhandare** [comment]
> @ashbhandare Thanks for the change, I just reviewed the SoftmaxCrossEntropyLoss code for adding MLFloat16 type, it looks fine but just for sanity sake I would quickly benchmark this kernel in isolation in a loop to make sure there are no perf regression due to casting that comes with this change.

I did the benchmark and saw no perf regression on the existing float test.

**Review by weixingzhang — APPROVED** [review]
